beacon_chain/state/helpers.py

```diff
@@ -1,72 +1,133 @@
 from beacon_chain.state.config import (
     DEFAULT_CONFIG,
 )
+from beacon_chain.state.shard_and_committee import (
+    ShardAndCommittee,
+)
+from beacon_chain.utils.blake import (
+    blake,
+)
+
+
+def is_power_of_two(num):
+    return ((num & (num - 1)) == 0) and num != 0
+
+
+def get_signed_parent_hashes(active_state,
+                             block,
+                             attestation,
+                             config=DEFAULT_CONFIG):
+    cycle_length = config['cycle_length']
+    oblique_parent_hashes = attestation.oblique_parent_hashes
 
+    parent_hashes = (
+        active_state.recent_block_hashes[
+            cycle_length + attestation.slot - block.slot_number:
+            cycle_length * 2 + attestation.slot - block.slot_number - len(oblique_parent_hashes)
+        ] +
+        oblique_parent_hashes
 
-def get_crosslink_shards_count(active_validators_count, config=DEFAULT_CONFIG):
-    """Returns how many shards in the `crosslink_shards` list
-    """
-    shard_count = config['shard_count']
-    crosslink_shards_count = (
-        active_validators_count // config['notaries_per_crosslink']
     )
-    if crosslink_shards_count > shard_count:
-        crosslink_shards_count = shard_count
-    return crosslink_shards_count
-
-
-def get_crosslink_shards(crystallized_state, config=DEFAULT_CONFIG):
-    """Returns a list of shards that will be crosslinking
-    """
-    shard_count = config['shard_count']
-    start_from = crystallized_state.next_shard
-    if start_from >= shard_count:
-        raise ValueError(
-            'start_from %d is larger than SHARD_COUNT %d',
-            start_from,
-            shard_count,
-        )
-
-    crosslink_shards_count = get_crosslink_shards_count(
-        crystallized_state.num_active_validators,
-        config=config,
+    return parent_hashes
+
+
+def get_attestation_indices(crystallized_state,
+                            attestation,
+                            config=DEFAULT_CONFIG):
+    last_state_recalc = crystallized_state.last_state_recalc
+    cycle_length = config['cycle_length']
+    indices_for_heights = crystallized_state.indices_for_heights
+
+    shard_position = list(filter(
+        lambda x: (
+            indices_for_heights[attestation.slot - last_state_recalc + cycle_length][x].shard_id ==
+            attestation.shard_id
+        ),
+        range(len(indices_for_heights[attestation.slot - last_state_recalc + cycle_length]))
+    ))[0]
+    attestation_indices = (
+        indices_for_heights[
+            attestation.slot - last_state_recalc + cycle_length
+        ][shard_position].committee
     )
 
-    if start_from + crosslink_shards_count <= shard_count:
-        return list(range(start_from, start_from + crosslink_shards_count))
+    return attestation_indices
```

Comment:
I maybe be completely off but does that corresponds to:

```Python
def get_indices_for_slot(crystallized_state, active_state, slot):
    lsr = crystallized_state.last_state_recalc
    assert lsr <= slot < lsr + CYCLE_LENGTH * 2
    return crystallized_state.indices_for_heights[slot - lsr]
```

Note that the spec has 2 errors on assert and "height"

```Python
def get_indices_for_slot(crystallized_state, active_state, slot):
    lsr = crystallized_state.last_state_recalc
    assert lsr <= slot < assert lsr + CYCLE_LENGTH * 2
    return crystallized_state.indices_for_heights[height - lsr]
```

If so, it seems very complex.

Reply:
This code was written before the `get_indices_for_slot` helper was added. This function actually serves this purpose from the spec

```
Let attestation_indices be get_indices_for_slot(crystallized_state, active_state, slot)[x], choosing x so that attestation_indices.shard_id equals the shard_id value provided to find the set of validators that is creating this attestation record
```

A lot of the complexity above is in the "choosing x" (or `shard_position`).

Will get those fixes in and also will get this helper function in to the codebase shortly. Looks like there might be a deeper bug and should be `slot - lsr + cycle_length` rather than just `slot - lsr`

Reply:
Yep, there was a bug. Please see the updated spec.